File: class07/python/TCC0701-Determinizacao.py
#!/usr/bin/env python3
"""TCC0701-Determinizacao.py | Aula 07
Algoritmo de construção de subconjuntos: AFN → AFD.
"""
from collections import defaultdict

# AFN exemplo: termina com abb
afn = defaultdict(set, {
    ("q0", "a"): {"q0", "q1"}, ("q0", "b"): {"q0"},
    ("q1", "b"): {"q2"},
    ("q2", "b"): {"q3"},
})
q0_afn = "q0"
finais_afn = {"q3"}
alfabeto = {"a", "b"}


def determinizar(afn_transicoes, q0, finais, alfabeto):
    """Constrói AFD equivalente por subconjuntos (BFS)."""
    afd_transicoes = {}
    afd_finais = set()
    estado_inicial = frozenset({q0})
    fila = [estado_inicial]
    visitados = {estado_inicial}

    while fila:
        C = fila.pop(0)
        if any(q in finais for q in C):
            afd_finais.add(C)
        for a in alfabeto:
            D = frozenset({p for q in C for p in afn_transicoes.get((q, a), set())})
            if not D:
                continue
            afd_transicoes[(C, a)] = D
            if D not in visitados:
                visitados.add(D)
                fila.append(D)

    return afd_transicoes, estado_inicial, afd_finais, visitados


if __name__ == "__main__":
    trans, q0_afd, finais_afd, estados = determinizar(afn, q0_afn, finais_afn, alfabeto)
    print(f"AFD: {len(estados)} estados, {len(finais_afd)} finais")
    for C in sorted(estados, key=lambda s: ("" if any(q in finais_afn for q in C) else "z", sorted(C))):
        final = " (final)" if C in finais_afd else ""
        print(f"  {set(C)}{final}")
    print("\nTransições:")
    for (C, a), D in sorted(trans.items(), key=lambda x: (sorted(x[0][0]), x[0][1])):
        print(f"  δ({set(C)}, {a}) = {set(D)}")

File: class07/java/TCC0701-Determinizacao.java
import java.util.*;

/**
 * TCC0701-Determinizacao.java | Aula 07
 * Algoritmo de construção de subconjuntos: AFN → AFD.
 * Mesma lógica do TCC0701-Determinizacao.py.
 */
class Determinizacao {

    record Par(String estado, String simbolo) {}

    static Map<Par, Set<String>> afn = Map.of(
            new Par("q0", "a"), Set.of("q0", "q1"),
            new Par("q0", "b"), Set.of("q0"),
            new Par("q1", "b"), Set.of("q2"),
            new Par("q2", "b"), Set.of("q3")
    );
    static String q0afn = "q0";
    static Set<String> finaisAfn = Set.of("q3");
    static Set<String> alfabeto = Set.of("a", "b");

    record AfdResult(Map<Set<String>, Map<String, Set<String>>> transicoes,
                     Set<String> q0, Set<Set<String>> finais, Set<Set<String>> estados) {}

    static AfdResult determinizar(Map<Par, Set<String>> afnTransicoes,
                                   String q0, Set<String> finais, Set<String> alfabeto) {
        Map<Set<String>, Map<String, Set<String>>> afdTrans = new LinkedHashMap<>();
        Set<Set<String>> afdFinais = new HashSet<>();
        Set<String> estadoInicial = Set.of(q0);
        Queue<Set<String>> fila = new LinkedList<>();
        Set<Set<String>> visitados = new LinkedHashSet<>();

        fila.add(estadoInicial);
        visitados.add(estadoInicial);

        while (!fila.isEmpty()) {
            Set<String> C = fila.poll();
            boolean ehFinal = C.stream().anyMatch(finais::contains);
            if (ehFinal) afdFinais.add(C);

            Map<String, Set<String>> transDeC = new LinkedHashMap<>();
            for (String a : alfabeto) {
                Set<String> D = new HashSet<>();
                for (String q : C) {
                    Set<String> dest = afnTransicoes.get(new Par(q, a));
                    if (dest != null) D.addAll(dest);
                }
                if (!D.isEmpty()) {
                    transDeC.put(a, D);
                    if (!visitados.contains(D)) {
                        visitados.add(D);
                        fila.add(D);
                    }
                }
            }
            afdTrans.put(C, transDeC);
        }
        return new AfdResult(afdTrans, estadoInicial, afdFinais, visitados);
    }

    public static void main(String[] args) {
        AfdResult r = determinizar(afn, q0afn, finaisAfn, alfabeto);
        System.out.println("AFD: " + r.estados().size() + " estados, " + r.finais().size() + " finais");
        for (Set<String> C : r.estados()) {
            String mark = r.finais().contains(C) ? " (final)" : "";
            System.out.println("  " + C + mark);
        }
        System.out.println("\nTransições:");
        for (var entry : r.transicoes().entrySet()) {
            for (var t : entry.getValue().entrySet()) {
                System.out.println("  δ(" + entry.getKey() + ", " + t.getKey() + ") = " + t.getValue());
            }
        }
    }
}
